src/overlays/actors/ovl_En_Dekubaba/z_en_dekubaba.c

```diff
@@ -44,188 +54,1325 @@ const ActorInit En_Dekubaba_InitVars = {
+            this->timer = 12;
+        }
+    } else if (this->timer == 18) {
+        Math_ScaledStepToS(&this->actor.shape.rot.x, 0x2AA8, 0xAAA);
+
+        if (Math_ScaledStepToS(&this->stemSectionAngle[0], 0x1554, 0x5B0)) {
+            this->timer = 25;
+        }
+
+        Math_ScaledStepToS(&this->stemSectionAngle[1], -0x38E3, 0xAAA);
+        Math_ScaledStepToS(&this->stemSectionAngle[2], -0x5C71, 0x2D8);
+    } else if (this->timer == 25) {
+        Math_ScaledStepToS(&this->actor.shape.rot.x, -0x5550, 0xAAA);
+
+        if (Math_ScaledStepToS(&this->stemSectionAngle[0], -0x6388, 0x93E)) {
+            this->timer = 26;
+        }
+
+        Math_ScaledStepToS(&this->stemSectionAngle[1], -0x3FFC, 0x4FA);
+        Math_ScaledStepToS(&this->stemSectionAngle[2], -0x238C, 0x444);
+    } else if (this->timer == 26) {
+        Math_ScaledStepToS(&this->actor.shape.rot.x, 0x1800, 0x93E);
+
+        if (Math_ScaledStepToS(&this->stemSectionAngle[0], -0x1555, 0x71C)) {
+            this->timer = 27;
+        }
+
+        Math_ScaledStepToS(&this->stemSectionAngle[1], -0x38E3, 0x2D8);
+        Math_ScaledStepToS(&this->stemSectionAngle[2], -0x5C71, 0x5B0);
+    } else if (this->timer >= 27) {
+        this->timer++;
+        if (this->timer > 30) {
+            if (this->actor.xzDistToPlayer < (80.0f * this->size)) {
+                EnDekubaba_SetupPrepareLunge(this);
+            } else {
+                EnDekubaba_SetupDecideLunge(this);
+            }
+        }
+    } else {
+        this->timer++;
+        if (this->timer == 10) {
+            Actor_PlaySfxAtPos(&this->actor, NA_SE_EN_DEKU_SCRAPE);
+        }
+
+        if (this->timer >= 12) {
+            Math_ScaledStepToS(&this->stemSectionAngle[2], -0x5C71, 0x88);
+        }
+    }
+
+    EnDekubaba_UpdateHeadPosition(this);
+}
+
+void EnDekubaba_SetupRecover(EnDekubaba* this) {
+    this->timer = 9;
+    this->collider.base.acFlags |= AC_ON;
+    this->skelAnime.playSpeed = -1.0f;
+    this->actionFunc = EnDekubaba_Recover;
+}
+
+void EnDekubaba_Recover(EnDekubaba* this, PlayState* play) {
+    s32 anyStepsDone;
+
+    SkelAnime_Update(&this->skelAnime);
+
+    if (this->timer > 8) {
+        anyStepsDone = Math_SmoothStepToS(&this->actor.shape.rot.x, 0x1800, 1, 0x11C6, 0x71C);
+        anyStepsDone |= Math_SmoothStepToS(&this->stemSectionAngle[0], -0x1555, 1, 0xAAA, 0x71C);
+        anyStepsDone |= Math_SmoothStepToS(&this->stemSectionAngle[1], -0x38E3, 1, 0xE38, 0x71C);
+        anyStepsDone |= Math_SmoothStepToS(&this->stemSectionAngle[2], -0x5C71, 1, 0x11C6, 0x71C);
+        if (!anyStepsDone) {
+            this->timer = 8;
+        }
+    } else {
+        if (this->timer != 0) {
+            this->timer--;
+        }
+
+        if (this->timer == 0) {
+            EnDekubaba_SetupDecideLunge(this);
+        }
+    }
+
+    EnDekubaba_UpdateHeadPosition(this);
+}
+
+void EnDekubaba_SetupHit(EnDekubaba* this, s32 arg1) {
```

Comment:
Any ideas for naming arg1? Looks like it sets the timer, but also determines type/color?

Reply:
I can't really think of a good name, it's not an effect type really because it doesn't correspond to anything. And yeah, the timer part doesn't help.

Reply:
Perhaps just `hitType`. Based on usage it does seem kind of consistent?

Reply:
Yeah we could do that. With an enum too? Might be a bit awkward.

Reply:
Yeah, there's an issue here that 0 and 1 are pretty much the same except for what they do to the timer.

Reply:
Actually I'm looking at the two functions together, what I hadn't realised was that `timer` is incremented in the individual functions rather than just in Update like normal. 
```c

void EnDekubaba_SetupHit(EnDekubaba* this, s32 hitType) {
    Animation_MorphToPlayOnce(&this->skelAnime, &gDekuBabaPauseChompAnim, -5.0f);
    this->timer = hitType;
    this->collider.base.acFlags &= ~AC_ON;
    Actor_SetScale(&this->actor, this->size * 0.01f);

    if (hitType == 2) {
        Actor_SetColorFilter(&this->actor, 0, 155, 0, 42);
    } else if (hitType == 3) {
        Actor_SetColorFilter(&this->actor, 0, 155, 0, 42);
        this->drawDmgEffType = ACTOR_DRAW_DMGEFF_ELECTRIC_SPARKS_LARGE;
        this->drawDmgEffScale = 0.75f;
        this->drawDmgEffAlpha = 2.0f;
    } else {
        Actor_SetColorFilter(&this->actor, 0x4000, 255, 0, 42);
    }
    this->actionFunc = EnDekubaba_Hit;
}

/**
 * Hit by a weapon or hit something when lunging.
 */
void EnDekubaba_Hit(EnDekubaba* this, PlayState* play) {
    s32 allStepsDone;

    SkelAnime_Update(&this->skelAnime);
    allStepsDone = true;
    allStepsDone &= Math_ScaledStepToS(&this->actor.shape.rot.x, -0x4000, 0xE38);
    allStepsDone &= Math_ScaledStepToS(this->stemSectionAngle, -0x4000, 0xE38);
    allStepsDone &= Math_ScaledStepToS(&this->stemSectionAngle[1], -0x4000, 0xE38);
    allStepsDone &= Math_ScaledStepToS(&this->stemSectionAngle[2], -0x4000, 0xE38);
    if (allStepsDone) {
        if (this->actor.colChkInfo.health == 0) {
            EnDekubaba_SetupShrinkDie(this);
        } else {
            this->collider.base.acFlags |= AC_ON;

            if (this->timer == 0) {
                if (this->actor.xzDistToPlayer < (80.0f * this->size)) {
                    EnDekubaba_SetupPrepareLunge(this);
                } else {
                    EnDekubaba_SetupRecover(this);
                }
            } else {
                EnDekubaba_SetupStunnedVertical(this);
            }
        }
    }

    EnDekubaba_UpdateHeadPosition(this);
}
```

This means it will stay on 0/1 until the head is in the right position, then it decides what to do partly based on the timer. So it is possible to distinguish the 0/not 0 cases, although it's still not completely cut-and-dried.

Reply:
Okay, think I got this one nailed down now.